File: 1539.kth-missing-positive-number.ts
/*
 * @lc app=leetcode id=1539 lang=typescript
 *
 * [1539] Kth Missing Positive Number
 */

// @lc code=start
function findKthPositive(arr: number[], k: number): number {
    let idx = 0;
    let left = 0;
    let right = arr.length - 1;

    while (left <= right) {
        const mid = Math.floor((left + right) / 2);
        if (arr[mid] - mid - 1 < k) {
            left = mid + 1;
        } else {
            right = mid - 1;
        }
    }

    return left + k;
}
// @lc code=end

File: 1539.kth-missing-positive-number.py
#
# @lc app=leetcode id=1539 lang=python3
#
# [1539] Kth Missing Positive Number
#

# @lc code=start
class Solution:
    def findKthPositive(self, arr: List[int], k: int) -> int:
        left, right = 0, len(arr) - 1

        while left <= right:
            mid = (left + right) // 2
            missing = arr[mid] - (mid + 1)

            if missing < k:
                left = mid + 1
            else:
                right = mid - 1
        
        return left + k
# @lc code=end 

